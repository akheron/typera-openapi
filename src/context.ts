import * as ts from 'typescript'

export type LogLevel = 'verbose' | 'info' | 'warn' | 'error'

export type Logger = (
  location: string,
  level: LogLevel,
  ...messages: any[]
) => void

export interface Context {
  checker: ts.TypeChecker
  source: ts.SourceFile
  log(level: LogLevel, ...messages: any[]): void
  location: ts.Node
}

export const context = (
  checker: ts.TypeChecker,
  source: ts.SourceFile,
  log: Logger,
  location: ts.Node
): Context => ({
  checker,
  source,
  log(level, ...messages) {
    const { line, character } = source.getLineAndCharacterOfPosition(
      // location is updated, get it from this
      this.location.getStart()
    )
    log(`${source.fileName}:${line + 1}:${character}`, level, ...messages)
  },
  location,
})

export const withLocation = (ctx: Context, location: ts.Node): Context => ({
  ...ctx,
  location,
})
