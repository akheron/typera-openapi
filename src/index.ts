export { generate } from './generate'
export { LogLevel } from './context'
import { OpenAPIV3 } from 'openapi-types'

export const prefix = (
  prefix: string,
  paths: OpenAPIV3.PathsObject
): OpenAPIV3.PathsObject =>
  Object.fromEntries(
    Object.entries(paths).map(([path, value]) => [prefix + path, value])
  )
