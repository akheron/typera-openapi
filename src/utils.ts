import * as ts from 'typescript'

export const isDefined = <T>(value: T | undefined): value is T =>
  value !== undefined

export const isOptional = (symbol: ts.Symbol): boolean =>
  !!(symbol.flags & ts.SymbolFlags.Optional)
export const isArrayType = (type: ts.Type): boolean =>
  type.symbol?.escapedName === 'Array'
export const isObjectType = (type: ts.Type): boolean =>
  !!(type.flags & ts.TypeFlags.Object)
export const isStringType = (type: ts.Type): boolean =>
  !!(type.flags & ts.TypeFlags.String)
export const isNumberType = (type: ts.Type): boolean =>
  !!(type.flags & ts.TypeFlags.Number)
export const isBooleanType = (type: ts.Type): boolean =>
  !!(type.flags & ts.TypeFlags.Boolean)
export const isBooleanLiteralType = (type: ts.Type): boolean =>
  !!(type.flags & ts.TypeFlags.BooleanLiteral)
export const isNumberLiteralType = (
  type: ts.Type
): type is ts.NumberLiteralType => !!(type.flags & ts.TypeFlags.NumberLiteral)
export const isStringLiteralType = (
  type: ts.Type
): type is ts.StringLiteralType => !!(type.flags & ts.TypeFlags.StringLiteral)
export const isUndefinedType = (type: ts.Type): boolean =>
  !!(type.flags & ts.TypeFlags.Undefined)
export const isNullType = (type: ts.Type): boolean =>
  !!(type.flags & ts.TypeFlags.Null)

export const getPropertyType = (
  checker: ts.TypeChecker,
  location: ts.Node,
  type: ts.Type,
  propertyName: string
): ts.Type | undefined => {
  const prop = type.getProperty(propertyName)
  if (!prop) return
  return checker.getTypeOfSymbolAtLocation(prop, location)
}

export const isPackageSymbol = (
  symbol: ts.Symbol,
  packageName: string,
  symbolName: string
): boolean => {
  const parent = (symbol.valueDeclaration ?? symbol.declarations[0]).parent
  return (
    symbol.name === symbolName &&
    ts.isSourceFile(parent) &&
    parent.fileName.includes(`/${packageName}/`)
  )
}
