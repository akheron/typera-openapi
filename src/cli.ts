#!/usr/bin/env node
import * as fs from 'fs'
import * as path from 'path'
import { generate } from '.'
import { Logger } from './context'
import { runPrettier } from './prettify'
import * as yargs from 'yargs'
import { OpenAPIV3 } from 'openapi-types'

const log: Logger = (location, level, ...args) =>
  console.log(`${location}: ${level}:`, ...args)

type Format = 'ts' | 'json'

const parseArgs = () =>
  yargs
    .usage('Usage: $0 [options] FILE...')
    .option('format', {
      description: 'Output file format',
      choices: ['ts' as const, 'json' as const],
      default: 'ts' as Format,
    })
    .option('prettify', {
      alias: 'p',
      description: 'Apply prettier to output TypeScript files',
      type: 'boolean',
      default: false,
    })
    .option('check', {
      alias: 'c',
      description:
        'Exit with an error if output files are not up-to-date (useful for CI)',
      type: 'boolean',
      default: false,
    }).argv

const outputFileName = (sourceFileName: string, ext: string): string =>
  sourceFileName.slice(0, -path.extname(sourceFileName).length) + ext

const main = async () => {
  const args = parseArgs()

  const sourceFiles = args._.map((x) => x.toString())
  const ext = `.openapi.${args.format}`

  const results = generate(sourceFiles, { strict: true }, { log }).map(
    (result) => ({
      ...result,
      outputFileName: outputFileName(result.fileName, ext),
    })
  )

  let success = true
  for (const { outputFileName, paths } of results) {
    let content = args.format === 'ts' ? tsString(paths) : jsonString(paths)
    if (args.prettify) {
      content = await runPrettier(outputFileName, content)
    }
    if (args.check) {
      if (!checkOutput(outputFileName, content)) success = false
    } else {
      writeOutput(outputFileName, content)
    }
  }

  if (!success) {
    process.exit(1)
  }
}

const checkOutput = (fileName: string, content: string): boolean => {
  const current = fs.readFileSync(fileName, 'utf-8')
  if (current !== content) {
    console.log(`${fileName} is out of date`)
    return false
  }
  return true
}

const writeOutput = (fileName: string, content: string): void => {
  console.log('Writing', fileName)
  fs.writeFileSync(fileName, content)
}

const tsString = (paths: OpenAPIV3.PathsObject): string => `\
import { OpenAPIV3 } from 'openapi-types'

const spec: { paths: OpenAPIV3.PathsObject } = ${JSON.stringify({ paths })};

export default spec;
`

const jsonString = (paths: OpenAPIV3.PathsObject): string =>
  JSON.stringify({ paths })

main()
