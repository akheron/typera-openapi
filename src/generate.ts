import * as ts from 'typescript'
import { OpenAPIV3 } from 'openapi-types'
import { Context, Logger, context, withLocation } from './context'
import {
  isDefined,
  isOptional,
  isObjectType,
  isUndefinedType,
  isPackageSymbol,
  getPropertyType,
  isNullType,
  isStringType,
  isNumberType,
  isBooleanType,
  isBooleanLiteralType,
  isNumberLiteralType,
  isStringLiteralType,
  isArrayType,
} from './utils'

interface GenerateOptions {
  log: Logger
}

interface Result {
  fileName: string
  paths: OpenAPIV3.PathsObject
}

export const generate = (
  fileNames: string[],
  compilerOptions: ts.CompilerOptions,
  options?: GenerateOptions
): Result[] => {
  const log = options?.log || (() => undefined)
  const program = ts.createProgram(fileNames, compilerOptions)
  const checker = program.getTypeChecker()

  const result: Result[] = []

  for (const sourceFile of program.getSourceFiles()) {
    if (!fileNames.includes(sourceFile.fileName)) continue
    if (sourceFile.isDeclarationFile) continue

    ts.forEachChild(sourceFile, (node) => {
      const paths = visit(context(checker, sourceFile, log, node), node)
      if (paths) {
        result.push({ fileName: sourceFile.fileName, paths })
      }
    })
  }

  return result
}

const visit = (
  ctx: Context,
  node: ts.Node
): OpenAPIV3.PathsObject | undefined => {
  if (ts.isExportAssignment(node) && !node.isExportEquals) {
    // 'export default' statement
    const argSymbols = getRouterCallArgSymbols(ctx, node.expression)
    if (!argSymbols) return

    const paths: OpenAPIV3.PathsObject = {}

    argSymbols.forEach((symbol) => {
      const location = symbol.valueDeclaration
      const routeDeclaration = getRouteDeclaration(
        withLocation(ctx, location),
        symbol
      )
      if (routeDeclaration) {
        const [path, pathItem] = routeDeclaration
        paths[path] = pathItem
      }
    })
    return paths
  }
}

const getRouterCallArgSymbols = (
  ctx: Context,
  expression: ts.Expression
): ts.Symbol[] | undefined => {
  if (!ts.isCallExpression(expression)) return

  const fn = expression.expression
  const args = expression.arguments

  // TODO: Check for router calls better than just checking the symbol name
  if (!ts.isIdentifier(fn) || fn.escapedText !== 'router') return

  const argSymbols = args
    .filter(ts.isIdentifier)
    .map((arg) => ctx.checker.getSymbolAtLocation(arg))
    .filter(isDefined)

  if (argSymbols.length !== args.length) return
  return argSymbols
}

const getRouteDeclaration = (
  ctx: Context,
  symbol: ts.Symbol
): [string, OpenAPIV3.PathItemObject] | undefined => {
  const description = getRouteDescription(ctx, symbol)
  const routeInput = getRouteInput(ctx, symbol)
  if (!routeInput) return
  const { method, path, requestNode, body, query, routeParams } = routeInput

  const responses = getResponseTypes(ctx, symbol)
  if (!responses) return

  const requestBody =
    requestNode && body
      ? typeToSchema(withLocation(ctx, requestNode), body)
      : undefined

  const parameters = [
    ...typeToParameters(ctx, 'path', routeParams),
    ...typeToParameters(ctx, 'query', query),
  ]

  const pathTemplate = path.replace(/:([^-.()/]+)\(.*?\)/g, '{$1}')

  return [
    pathTemplate,
    {
      [method]: {
        ...(description ? { description } : undefined),
        ...(parameters.length > 0 ? { parameters } : undefined),
        ...operationRequestBody(requestBody),
        responses,
      },
    },
  ]
}

const getRouteDescription = (ctx: Context, symbol: ts.Symbol) =>
  symbol
    .getDocumentationComment(ctx.checker)
    .map((part) => part.text)
    .join('')

const operationRequestBody = (
  contentSchema: OpenAPIV3.SchemaObject | undefined
): { requestBody: OpenAPIV3.RequestBodyObject } | undefined => {
  if (!contentSchema) return
  return {
    requestBody: { content: { 'application/json': { schema: contentSchema } } },
  }
}

const methodNames = [
  'get',
  'post',
  'put',
  'delete',
  'head',
  'options',
  'patch',
  'all',
]

interface RouteInput {
  method: string
  path: string
  requestNode: ts.Node | undefined
  body: ts.Type | undefined
  query: ts.Type | undefined
  routeParams: ts.Type | undefined
}

const getRouteInput = (
  ctx: Context,
  symbol: ts.Symbol
): RouteInput | undefined => {
  const declaration = symbol.valueDeclaration
  if (!ts.isVariableDeclaration(declaration)) return

  let expr = declaration.initializer
  if (!expr) return

  let method: string | undefined,
    path: string | undefined,
    requestNode: ts.Node | undefined,
    body: ts.Type | undefined,
    query: ts.Type | undefined,
    routeParams: ts.Type | undefined

  while (ts.isCallExpression(expr)) {
    const lhs: ts.LeftHandSideExpression = expr.expression
    if (ts.isIdentifier(lhs)) {
      // Direct route() call
      if (expr.arguments.length != 2) {
        ctx.log('warn', 'Expected 2 arguments for route()')
        return
      }
      const methodArg = expr.arguments[0]
      const pathArg = expr.arguments[1]
      if (!ts.isStringLiteral(methodArg) || !ts.isStringLiteral(pathArg)) {
        ctx.log('warn', 'method and path must be string literals')
        return
      }
      method = methodArg.text
      path = pathArg.text

      // Property access chain ends
      break
    } else if (ts.isPropertyAccessExpression(lhs)) {
      if (!ts.isIdentifier(lhs.name)) return
      const fnName = lhs.name.escapedText.toString()
      if (methodNames.includes(fnName)) {
        method = fnName
        const pathArg = expr.arguments[0]
        if (!pathArg) {
          ctx.log('warn', `No argument for method ${fnName}`)
          return
        }
        if (!ts.isStringLiteral(pathArg)) {
          ctx.log('warn', `Method ${fnName} argument is not a string literal`)
          return
        }
        path = pathArg.text
      } else if (fnName === 'handler') {
        const handlerFn = expr.arguments[0]
        if (
          !handlerFn ||
          (!ts.isArrowFunction(handlerFn) &&
            !ts.isFunctionExpression(handlerFn))
        ) {
          ctx.log('warn', 'Handler is not a function')
          return
        }
        const req = handlerFn.parameters[0]
        if (req) {
          const type = ctx.checker.getTypeAtLocation(req)
          body = getPropertyType(ctx.checker, req, type, 'body')
          query = getPropertyType(ctx.checker, req, type, 'query')
          routeParams = getPropertyType(ctx.checker, req, type, 'routeParams')
          requestNode = req
        }
      }
      expr = lhs.expression
    } else {
      // Do nothing, I guess
    }
  }
  if (!method) {
    ctx.log('warn', 'Could not determine route method')
    return
  }
  if (!path) {
    ctx.log('warn', 'Could not determine route path')
    return
  }
  if (method === 'all') {
    ctx.log('warn', "The 'all' method is not supported, skipping route")
    return
  }
  return { method, path, requestNode, body, query, routeParams }
}

const getResponseTypes = (
  ctx: Context,
  symbol: ts.Symbol
): OpenAPIV3.ResponsesObject | undefined => {
  const descriptions = getResponseDescriptions(symbol)

  const routeType = ctx.checker.getTypeOfSymbolAtLocation(
    symbol,
    symbol.valueDeclaration
  )

  if (
    !routeType.aliasSymbol ||
    !isPackageSymbol(routeType.aliasSymbol, 'typera-common', 'Route')
  ) {
    return
  }

  const args = routeType.aliasTypeArguments
  if (!args || args.length !== 1) {
    throw new Error('expected 1 type argument for Route')
  }
  const responseType = args[0]

  const result: OpenAPIV3.ResponsesObject = {}
  if (isObjectType(responseType)) {
    const responseDef = getResponseDefinition(ctx, descriptions, responseType)
    if (responseDef) result[responseDef.status] = responseDef.response
  } else if (responseType.isUnion()) {
    responseType.types.forEach((type) => {
      const responseDef = getResponseDefinition(ctx, descriptions, type)
      if (responseDef) result[responseDef.status] = responseDef.response
    })
  }

  if (Object.keys(result).length === 0) {
    // Any response types could not be determined so this is not a valid route after all
    return
  }

  return result
}

const getResponseDescriptions = (
  symbol: ts.Symbol
): Partial<Record<string, string>> =>
  Object.fromEntries(
    symbol
      .getJsDocTags()
      .filter((tag) => tag.name === 'response')
      .map((tag) => tag.text)
      .filter(isDefined)
      .map((text) => {
        const match = /(\d{3}) (.+)/.exec(text)
        if (!match) return undefined
        return [match[1], match[2]] as const
      })
      .filter(isDefined)
  )

const getResponseDefinition = (
  ctx: Context,
  descriptions: Partial<Record<string, string>>,
  responseType: ts.Type
): { status: string; response: OpenAPIV3.ResponseObject } | undefined => {
  const statusSymbol = responseType.getProperty('status')
  const bodySymbol = responseType.getProperty('body')
  const headersSymbol = responseType.getProperty('headers')
  if (!statusSymbol || !bodySymbol || !headersSymbol) return

  const statusType = ctx.checker.getTypeOfSymbolAtLocation(
    statusSymbol,
    statusSymbol.valueDeclaration
  )
  const bodyType = ctx.checker.getTypeOfSymbolAtLocation(
    bodySymbol,
    bodySymbol.valueDeclaration
  )
  const headersType = ctx.checker.getTypeOfSymbolAtLocation(
    headersSymbol,
    headersSymbol.valueDeclaration
  )
  if (!statusType || !bodyType || !headersType) return

  if (!(statusType.flags & ts.TypeFlags.NumberLiteral)) {
    return
  }

  const status = ctx.checker.typeToString(statusType)

  // TODO: If bodyType is an interface (or type alias?), generate a schema
  // component object and a reference to it?
  let bodySchema: OpenAPIV3.SchemaObject | undefined
  if (!isUndefinedType(bodyType)) {
    bodySchema = typeToSchema(ctx, bodyType)
    if (!bodySchema) return
  }

  const headers = !isUndefinedType(headersType)
    ? typeToHeaders(ctx, headersType)
    : undefined

  let description = descriptions[status]
  if (!description) {
    ctx.log('warn', `No description for response ${status}`)
    description = status
  }

  return {
    status,
    response: {
      description,
      ...(bodySchema
        ? {
            content:
              isStringType(bodyType) || isNumberType(bodyType)
                ? { 'text/plain': { schema: bodySchema } }
                : {
                    'application/json': {
                      schema: bodySchema,
                    },
                  },
          }
        : undefined),
      ...(headers
        ? {
            headers,
          }
        : undefined),
    },
  }
}

const typeToParameters = (
  ctx: Context,
  in_: 'path' | 'query' | 'header' | 'cookie',
  type: ts.Type | undefined
): OpenAPIV3.ParameterObject[] => {
  if (!type) return []

  const props = ctx.checker.getPropertiesOfType(type)
  return props.map((prop) => ({
    name: prop.name,
    in: in_,
    required: in_ === 'path' ? true : !isOptional(prop),
  }))
}

interface Headers {
  [header: string]: OpenAPIV3.HeaderObject
}

const typeToHeaders = (ctx: Context, type: ts.Type): Headers => {
  const result: Headers = {}
  const props = ctx.checker.getPropertiesOfType(type)
  props.forEach((prop) => {
    result[prop.name] = {
      required: !isOptional(prop),
    }
  })
  return result
}

const typeToSchema = (
  ctx: Context,
  type: ts.Type,
  optional = false
): OpenAPIV3.SchemaObject | undefined => {
  let nullable: { nullable: true } | undefined

  if (type.isUnion()) {
    let elems = type.types

    if (optional) {
      elems = type.types.filter((elem) => !isUndefinedType(elem))
    }

    if (elems.some(isNullType)) {
      // One of the union elements is null
      nullable = { nullable: true }
      elems = elems.filter((elem) => !isNullType(elem))
    }

    if (elems.every(isBooleanLiteralType)) {
      // All elements are boolean literals => boolean
      return { type: 'boolean', ...nullable }
    } else if (elems.every(isNumberLiteralType)) {
      // All elements are number literals => enum
      return {
        type: 'number',
        enum: elems.map((elem) => elem.value),
        ...nullable,
      }
    } else if (elems.every(isStringLiteralType)) {
      // All elements are string literals => enum
      return {
        type: 'string',
        enum: elems.map((elem) => elem.value),
        ...nullable,
      }
    } else if (elems.length >= 2) {
      // 2 or more types remain => anyOf
      return {
        anyOf: elems.map((elem) => typeToSchema(ctx, elem)).filter(isDefined),
        ...nullable,
      }
    } else {
      // Only one element left in the union. Fall through and consider it as the
      // sole type.
      type = elems[0]
    }
  }

  if (isArrayType(type)) {
    const elemType = type.getNumberIndexType()
    if (!elemType) {
      ctx.log('warn', 'Could not get array element type')
      return
    }
    const elemSchema = typeToSchema(ctx, elemType)
    if (!elemSchema) return

    return { type: 'array', items: elemSchema, ...nullable }
  }

  if (
    isObjectType(type) ||
    (type.isIntersection() && type.types.every((part) => isObjectType(part)))
  ) {
    const props = ctx.checker.getPropertiesOfType(type)
    return {
      type: 'object',
      required: props
        .filter((prop) => !isOptional(prop))
        .map((prop) => prop.name),
      ...nullable,
      properties: Object.fromEntries(
        props
          .map((prop) => {
            const propType = ctx.checker.getTypeOfSymbolAtLocation(
              prop,
              ctx.location
            )
            if (!propType) {
              ctx.log('warn', 'Could not get type for property', prop.name)
              return
            }
            const propSchema = typeToSchema(ctx, propType, isOptional(prop))
            if (!propSchema) {
              ctx.log('warn', 'Could not get schema for property', prop.name)
              return
            }
            return [prop.name, propSchema]
          })
          .filter(isDefined)
      ),
    }
  }

  if (isStringType(type)) {
    return { type: 'string', ...nullable }
  }
  if (isNumberType(type)) {
    return { type: 'number', ...nullable }
  }
  if (isBooleanType(type)) {
    return { type: 'boolean', ...nullable }
  }
  if (isStringLiteralType(type)) {
    return { type: 'string', enum: [type.value], ...nullable }
  }
  if (isNumberLiteralType(type)) {
    return { type: 'number', enum: [type.value], ...nullable }
  }

  const typeStr = ctx.checker.typeToString(type)
  if (typeStr === 'Branded<number, IntBrand>') {
    // io-ts int
    return { type: 'integer', ...nullable }
  }

  ctx.log('warn', `Ignoring an unknown type: ${typeStr}`)
  return
}
