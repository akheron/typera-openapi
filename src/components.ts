import * as ts from 'typescript'
import { OpenAPIV3 } from 'openapi-types'
import { isInterface, isTypeAlias } from './utils'

export class Components {
  // undefined acts as a placeholder
  #schemas: Map<string, OpenAPIV3.SchemaObject | undefined>
  #symbolSchemas: Map<ts.Symbol, string>

  constructor() {
    this.#schemas = new Map()
    this.#symbolSchemas = new Map()
  }

  withSymbol(
    symbol: ts.Symbol | undefined,
    run: (
      addComponent: () => void
    ) => OpenAPIV3.SchemaObject | OpenAPIV3.ReferenceObject | undefined
  ): OpenAPIV3.SchemaObject | OpenAPIV3.ReferenceObject | undefined {
    const ref = symbol && this.#getRefForSymbol(symbol)
    if (ref) {
      return { $ref: ref }
    }

    if (symbol && (isInterface(symbol) || isTypeAlias(symbol))) {
      let added = false
      const schema = run(() => {
        this.#addSymbol(symbol)
        added = true
      })
      if (added) {
        if (schema === undefined || '$ref' in schema) {
          this.#deleteSymbol(symbol)
          return schema
        } else {
          return { $ref: this.#addSchema(symbol, schema) }
        }
      } else {
        return schema
      }
    } else {
      return run(() => {})
    }
  }

  #getRefForSymbol(symbol: ts.Symbol): string | undefined {
    const schemaName = this.#symbolSchemas.get(symbol)
    return schemaName !== undefined
      ? `#/components/schemas/${schemaName}`
      : undefined
  }

  #addSymbol(symbol: ts.Symbol) {
    if (this.#symbolSchemas.has(symbol)) return

    let name = symbol.name
    for (let i = 2; ; i++) {
      if (this.#schemas.has(name)) {
        name = `${symbol.name}${i}`
      } else {
        break
      }
    }

    this.#schemas.set(name, undefined)
    this.#symbolSchemas.set(symbol, name)
  }

  #deleteSymbol(symbol: ts.Symbol) {
    const name = this.#symbolSchemas.get(symbol)
    if (name === undefined) return

    this.#schemas.delete(name)
    this.#symbolSchemas.delete(symbol)
  }

  #addSchema(symbol: ts.Symbol, schema: OpenAPIV3.SchemaObject): string {
    const name = this.#symbolSchemas.get(symbol)
    if (name === undefined)
      throw new Error(`No schema has been added for symbol ${symbol.name}`)
    this.#schemas.set(name, schema)
    return `#/components/schemas/${name}`
  }

  build(): OpenAPIV3.ComponentsObject {
    const schemas = Object.fromEntries(
      [...this.#schemas.entries()].flatMap(([k, v]) =>
        v !== undefined ? [[k, v]] : []
      )
    )
    return {
      ...(this.#schemas.size > 0 ? { schemas } : undefined),
    }
  }
}
