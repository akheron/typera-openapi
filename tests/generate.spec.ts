import * as path from 'path'
import { LogLevel, generate } from '../src/index'

const relativePath = (fileName: string): string =>
  path.relative(process.cwd(), __dirname + '/' + fileName)

const testRoutes = relativePath('test-routes.ts')
const warningRoutes = relativePath('warning-routes.ts')

describe('generate', () => {
  it('works', () => {
    const result = generate([testRoutes], { strict: true })
    expect(result).toMatchSnapshot()
  })

  it('warnings', () => {
    interface LogMessage {
      location: string
      level: LogLevel
      message: string
    }
    const warnings: LogMessage[] = []
    generate(
      [warningRoutes],
      { strict: true },
      {
        log: (location, level, ...messages) =>
          warnings.push({ location, level, message: messages.join(' ') }),
      }
    )
    expect(warnings).toMatchSnapshot()
  })
})
