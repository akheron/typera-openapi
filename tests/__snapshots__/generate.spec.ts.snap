// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate warnings 1`] = `
Array [
  Object {
    "level": "warn",
    "location": "tests/warning-routes.ts:3:6",
    "message": "No description for response 200",
  },
]
`;

exports[`generate works 1`] = `
Array [
  Object {
    "fileName": "tests/test-routes.ts",
    "paths": Object {
      "/branded-request-body": Object {
        "post": Object {
          "requestBody": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "param": Object {
                      "type": "integer",
                    },
                  },
                  "required": Array [
                    "param",
                  ],
                  "type": "object",
                },
              },
            },
          },
          "responses": Object {
            "200": Object {
              "content": Object {
                "text/plain": Object {
                  "schema": Object {
                    "type": "number",
                  },
                },
              },
              "description": "200",
            },
            "400": Object {
              "content": Object {
                "text/plain": Object {
                  "schema": Object {
                    "type": "string",
                  },
                },
              },
              "description": "400",
            },
          },
        },
      },
      "/constant": Object {
        "get": Object {
          "description": "No input, static output",
          "responses": Object {
            "200": Object {
              "content": Object {
                "text/plain": Object {
                  "schema": Object {
                    "type": "string",
                  },
                },
              },
              "description": "Successful result",
            },
          },
          "summary": "This is a summary",
          "tags": Array [
            "Tag",
          ],
        },
      },
      "/direct-route-call": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "content": Object {
                "text/plain": Object {
                  "schema": Object {
                    "type": "string",
                  },
                },
              },
              "description": "200",
            },
          },
        },
      },
      "/interface-array-response": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "content": Object {
                "application/json": Object {
                  "schema": Object {
                    "items": Object {
                      "properties": Object {
                        "petName": Object {
                          "nullable": true,
                          "type": "string",
                        },
                        "shoeSize": Object {
                          "type": "number",
                        },
                      },
                      "required": Array [
                        "shoeSize",
                        "petName",
                      ],
                      "type": "object",
                    },
                    "type": "array",
                  },
                },
              },
              "description": "200",
            },
          },
        },
      },
      "/interface-response": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "content": Object {
                "application/json": Object {
                  "schema": Object {
                    "properties": Object {
                      "petName": Object {
                        "nullable": true,
                        "type": "string",
                      },
                      "shoeSize": Object {
                        "type": "number",
                      },
                    },
                    "required": Array [
                      "shoeSize",
                      "petName",
                    ],
                    "type": "object",
                  },
                },
              },
              "description": "200",
            },
          },
        },
      },
      "/no-explicit-route-type": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "content": Object {
                "text/plain": Object {
                  "schema": Object {
                    "type": "string",
                  },
                },
              },
              "description": "200",
            },
          },
        },
      },
      "/query": Object {
        "get": Object {
          "parameters": Array [
            Object {
              "in": "query",
              "name": "str",
              "required": true,
            },
            Object {
              "in": "query",
              "name": "num",
              "required": false,
            },
          ],
          "responses": Object {
            "200": Object {
              "content": Object {
                "text/plain": Object {
                  "schema": Object {
                    "type": "string",
                  },
                },
              },
              "description": "200",
            },
            "400": Object {
              "content": Object {
                "text/plain": Object {
                  "schema": Object {
                    "type": "string",
                  },
                },
              },
              "description": "400",
            },
          },
        },
      },
      "/request-body": Object {
        "post": Object {
          "description": "This one has request body and two possible successful responses",
          "requestBody": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "nullableNum": Object {
                      "nullable": true,
                      "type": "number",
                    },
                    "nullableObj": Object {
                      "nullable": true,
                      "properties": Object {
                        "foo": Object {
                          "type": "number",
                        },
                      },
                      "required": Array [
                        "foo",
                      ],
                      "type": "object",
                    },
                    "numLit": Object {
                      "enum": Array [
                        42,
                      ],
                      "type": "number",
                    },
                    "numLits": Object {
                      "enum": Array [
                        42,
                        123,
                      ],
                      "type": "number",
                    },
                    "optionalBool": Object {
                      "type": "boolean",
                    },
                    "requiredBool": Object {
                      "type": "boolean",
                    },
                    "str": Object {
                      "type": "string",
                    },
                    "strLit": Object {
                      "enum": Array [
                        "foo",
                      ],
                      "type": "string",
                    },
                    "strLits": Object {
                      "enum": Array [
                        "foo",
                        "bar",
                      ],
                      "type": "string",
                    },
                  },
                  "required": Array [
                    "str",
                    "requiredBool",
                    "nullableNum",
                    "nullableObj",
                    "numLit",
                    "numLits",
                    "strLit",
                    "strLits",
                  ],
                  "type": "object",
                },
              },
            },
          },
          "responses": Object {
            "200": Object {
              "content": Object {
                "text/plain": Object {
                  "schema": Object {
                    "type": "string",
                  },
                },
              },
              "description": "Successful result",
            },
            "201": Object {
              "description": "A new resource was created",
            },
            "400": Object {
              "content": Object {
                "text/plain": Object {
                  "schema": Object {
                    "type": "string",
                  },
                },
              },
              "description": "Validation error",
            },
          },
        },
      },
      "/response-headers": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "content": Object {
                "text/plain": Object {
                  "schema": Object {
                    "type": "string",
                  },
                },
              },
              "description": "200",
              "headers": Object {
                "X-Bar": Object {
                  "required": false,
                },
                "X-Foo": Object {
                  "required": true,
                },
              },
            },
          },
        },
      },
      "/unused-request": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "content": Object {
                "text/plain": Object {
                  "schema": Object {
                    "type": "string",
                  },
                },
              },
              "description": "200",
            },
          },
        },
      },
      "/user/{id}": Object {
        "get": Object {
          "parameters": Array [
            Object {
              "in": "path",
              "name": "id",
              "required": true,
            },
          ],
          "responses": Object {
            "200": Object {
              "content": Object {
                "application/json": Object {
                  "schema": Object {
                    "properties": Object {
                      "id": Object {
                        "type": "number",
                      },
                    },
                    "required": Array [
                      "id",
                    ],
                    "type": "object",
                  },
                },
              },
              "description": "200",
            },
          },
        },
      },
    },
  },
]
`;
